### PR TITLE
Staking module rebuild (Track F, F1–F10)

### PR DESCRIPTION
**Pull request by codewiththeflowxyz**
**Full Track F Staking module rebuild for the Webapp V2 redesign.** Base: `app-redesign`. Ten tickets (F1–F10), each landed as a linear slice on this branch; F10 closes the track with a module-wide QA + design-review + curated-e2e round.

## Done
- **F1 · calldata seam (APP-337):** extracted pure `generateStakeCalldata` + `calculateStakeApprovalAmounts` and the `useStakeCalldata` hook wrapper, with a combinatorial golden-master matrix proving parity against the legacy context (no UI).
- **F2 · destination page shell (APP-338):** dev-only `/stake-v2` route with URL-synced **My positions / Statistics / About** tabs. Statistics: Rate/TVL chart, details strip, engine promo card, borrow-utilization block. About: corpus copy, how-it-works, links.
- **F3 · My positions tab (APP-339):** positions table (per-urn staked/borrowed, mini risk meter, claimable rewards, hide-inactive toggle), aggregate summary card (totals, claimable/earned rewards, signed Net APY), activity table (subgraph events grouped per tx into the combined verb taxonomy, per-position filter). Borrowed figures use live Vat debt; reward stats show the actually-claimable token icons.
- **F4 · open-position takeover (APP-311):** full-screen open flow — stake amount, optional borrow card (risk slider bounded by dust / debt-ceiling / capped safe max), delegate selection, confirm summary, EIP-5792 batched multicall through the F1 seam.
- **F5 · position management:** position details modal + "Manage a position" sheet (stake/withdraw, borrow/repay with the dust-gap max-repay rule, delegate change), menu deep links pre-toggling the matching cards, all through the same launch/calldata seam.
- **F6 · claim flows & inactive states:** claim-rewards modal (multi-reward selection, claim & restake), inactive-position treatment and the reopen takeover converging on the open flow.
- **F7 · route flip (APP-341):** V2 now serves at `/stake`; `/stake-v2` and the legacy `StakePanes`/`StakeModuleWidget` UI are deleted.
- **F8 · liquidation edge states (APP-313):** at-risk warning banner with Stake SKY / Repay debt deep links, liquidated badge + past-tense banner with Claim CTA, liquidation post-mortem modal (fields the indexer can't provide render an honest `–`), bundled claim+withdraw recovery via EIP-5792. Includes the risk-math fix: proximity short-circuits to 100 whenever the liquidation price ≥ the OSM/delayed price, so a recovering market price can never mask on-chain liquidatability. Exercised end-to-end against real `dog.bark`/`clip.take` liquidations on a fork.
- **F9 · data-point audit (APP-342):** committed inventory at `apps/webapp/src/modules/stake/DATA-POINTS.md` — every displayed value on every surface traced to its exact source (naming which price feeds each USD conversion), transform/units, formatter and refresh semantics, plus a cross-surface consistency map and a live fork reconciliation (28/28 checks). Fixes landed with tests:
  - post-tx invalidation now covers wagmi's batched `readContracts` and `simulateDrip` keys — claimables, wallet balances and live total debt refresh after every stake tx (previously silently stale; proven live with a real claim);
  - failed reads render `–` instead of healthy-looking values (unlit risk meter, `$0.00` claimables);
  - open-flow risk slider clamps to the OSM-capped max it prints as its own ceiling;
  - the details-modal claim chip pairs its amount with a single reward token (no cross-token sums);
  - both Statistics "Current Rate" stats now agree — the chart hero is sourced from the on-chain stability fee;
  - price displays pinned to 4 decimals; loading-gate and zero-format alignments.
  Remaining behavioral findings are catalogued as AUD-1…18 in the inventory's register with per-item follow-up scopes.
- **F10 · deep QA + design review + curated e2e (APP-343):** committed companion at `apps/webapp/src/modules/stake/QA-CASES.md` — three verification axes:
  - **Figma coverage sweep** of both design files, page by page: 43 stake frames mapped implemented / partial / missing / deferred. One designed feature is missing (open-flow reward-token selector → AUD-19), one deliberately deferred (details-menu "Change reward"/"Close position" → APP-312), one systematically simplified (designed slider ticks/markers/gradient — vendored shadcn Slider). No mobile stake designs exist (TBD in hi-fi).
  - **54-case QA matrix** exercised on a live fork (real underwater urn, real `bark`/`take` liquidation): 53 pass, 0 functional failures, 1 n/e (wallet rejection — the mock wallet auto-approves; unit-covered).
  - **Curated e2e suite** `src/test/e2e/tests/stake-onchain.spec.ts`: 8 specs — 2 read smokes + exactly one spec per contract-write path (open+borrow+delegate multicall · stake/withdraw · borrow/dust-guard/wipe-all · delegate change · claim + claim-&-restake · liquidation recovery bundle staged via a real on-chain bark+auction). Every write is asserted against `vat.urns` ink/art, farm `earned()` and ERC-20 balances — never success copy, since the mock wallet's `wallet_sendCalls` is non-atomic and reports optimistically. Promotion/rejection rationale per case in QA-CASES.md §3; ~2–3 min wall-clock.
  - **Design review vs hi-fi/UX frames** (31 findings, evidence pairs archived): trivial objective fixes landed in-PR with tests — visible slider tracks (shadcn's `bg-muted` resolves to transparent; `--color-muted` is undefined in the theme), hi-fi risk-meter geometry, persistent details-menu chevrons, "30,000 USDS" spelled out in the dust callout, and the dust-floor borrow error formatted module-side. Judgment-level findings are documented for design decisions, not coded.

## Gates & follow-ups
- **Merge gate:** liquidation surfaces (F8) are built to the UX frames — hi-fi sign-off pending; the F10 review of those surfaces is structural-only until hi-fi lands. The at-risk warning threshold is a flagged placeholder (proximity ≥ 40) awaiting the product number.
- **Follow-up ticket candidates from F10:** open-flow reward-token selector (AUD-19 — designed, SPK farm already live so the multi-reward condition holds); out-of-range `urn_index` renders a functional phantom-position modal (AUD-20); activity Status column hardcodes `Completed` where the design draws `Pending` (AUD-21); info-tooltips missing on details-modal stats.
- **Design-system-level decisions surfaced by the review** (report, not code): legacy token set vs the redesign palette/type ramp (one mapping decision covers most color/type deltas); risk-meter palette incl. a raw `bg-orange-400` literal shared app-wide; liquidation-zone indicator draws semantic thresholds (deliberate F5 behavior) where the design draws visual quarters; the confirm step-list treatment is a component build on the shared `TransactionModal`; a tarmac-owned slider variant is needed for the designed ticks/markers/gradient.
- Error-state design pass (register AUD-5 residual) and the price-source labeling decision (AUD-3 / UI-spec open question §5) remain follow-ups, not blockers.
- One engine one-liner deliberately left out pending owner sign-off: `useCollateralData` scope-key typo (AUD-14).
- Indexer ticket candidates: bark/auction fields for the post-mortem's `–` gaps; price-at-event for historical USD valuations.

## Notes
- About copy stays corpus-fed; activity rows are confirmed-only (optimistic pending rows remain an open product decision); no Stake mobile designs yet.
- Loops at the F10 tip: unit 1817/1817, hooks 495/495, typecheck/lint/prettier green; curated stake e2e 8/8 on a fresh vnet.

### CODE REVIEW

**Comment by iceanddust** [comment]
@coderabbitai review

**Comment by iceanddust** [comment]
### Code review

Found 1 issue:

1. Borrow validity uses strict `<` against the debt-ceiling headroom, but the 100% chip stages exactly that boundary value. When the debt ceiling (not collateral) is the binding constraint, `maxBorrowable` equals `availableBorrowFromDebtCeiling`, so tapping 100% makes `borrowValid` false at equality while the error check uses `>` — Confirm is silently disabled with no error message at the max the UI itself advertised. Same defect in both flows; fix with `<=` on the ceiling bound (or stage max − 1 wei from the chip).

https://github.com/jetstreamgg/tarmac/blob/fc15565c4dd566a5c23726d16d80e77ef26283a6/apps/webapp/src/modules/stake/components/OpenPositionTakeover.tsx#L171-L176

https://github.com/jetstreamgg/tarmac/blob/fc15565c4dd566a5c23726d16d80e77ef26283a6/apps/webapp/src/modules/stake/components/ManagePositionTakeover.tsx#L201-L203

🤖 Generated with [Claude Code](https://claude.ai/code)

<sub>- If this code review was useful, please react with 👍. Otherwise, react with 👎.</sub>

**Review by iceanddust — APPROVED** [review]
Nice job! This module has always been the largest and most complex one, I did some testing and it seems to work pretty well! While testing, I noticed some minor stuff which I'll post below, but I think these could go as a follow-up PR since I think the scope of this PR is already met: 
- When the user doesn't have any open positions and for the unconnected state I think we should default to the Statistics tab, that way the first thing the user sees is not an empty state page
- The risk pill component needs to be updated to reuse the one in the earn and portfolio pages
- The rate chart and data points are currently showing the Borrow Rate, but should they instead display the Rewards Rate as that's what the user can earn? Probably a question for Kacper
- In the borrow utilization progress bar, it seems like the legend points are missing the colors (purple for the filled portion, grayish for the empty portion)
- I think we should display in the transaction overview that we're automatically selecting the SKY reward for the user's position
- After supplying more SKY, the position on the left and the total supplied on the right updated successfully, but when I tried to withdraw and borrow, the values didn't update after the successful transaction
- It looks like we will now support mixed flows like Supply and repay or Withdraw and borrow. We should make sure those work properly (perhaps with the e2e tests when the time comes) since those flows weren't conteplated in the previous version of the app. I did some small testing and they are working well, but there could be edge cases. Also, the borrow limits and progress bar may get affected by this

**Comment by codewiththeflowxyz** [comment]
Thanks for the review and the testing round — all seven points are addressed in d902cb95a, point by point:

1. **Default tab for unconnected / no-positions states** — done. With no explicit `tab=` param the page now lands on Statistics when disconnected or when the positions query settles empty; `?tab=` always wins, and clicking any tab (even the active one) pins it into the URL so a late empty-settle can never switch the view away from under the user.
2. **Risk pill reuse** — done. `RiskMeter` is now a purely presentational pill (per-segment fill classes); the earn/portfolio/product-detail surfaces use a `RiskTierMeter` wrapper for the product-risk mapping, and the stake positions table maps liquidation proximity onto the same shared pill. One component app-wide, domain mappings outside it.
3. **Rate chart: Borrow Rate vs Rewards Rate** — left as an open product question, no code change; flagging it for a product decision.
4. **Borrow-utilization legend colors** — done. The bar fill and the Borrowed dot now use the hi-fi purple (`#757dff`, same raw-value precedent as the rate chart until the token-mapping decision lands); Available stays muted gray.
5. **Auto-selected reward in the transaction overview** — done. The confirm summary shows "Reward (selected automatically)" with the selected farm's token, resolved via a pure address-book lookup; the stake card's reward symbol now also derives from the actual farm (matters for reopening an urn attached to a non-SKY farm).
6. **Values not refreshing after withdraw + borrow** — root-caused and fixed. Invalidation fires at mined success, but the single subgraph refetch could land before the indexer had the mutation and re-cache pre-tx rows. All four tx success paths now share one `invalidateStakeQueries` helper: on-chain reads refetch once, subgraph keys re-invalidate on a 5s/15s trail to outwait indexer lag.
7. **Mixed flows** — covered. Two new on-chain e2e specs exercise withdraw+borrow and supply+repay as single bundles, asserting exact `vat.urns` ink/art deltas and token balances; the curated suite is now 10 specs.

While validating the suite on fresh forks, two more fixes landed in the same commit: the "Liquidation price is higher than the capped OSM SKY price" withdraw error had become unreachable (always shadowed by the generic risk error since the proximity short-circuit) and now takes precedence; and the stake e2e specs are hardened against live OSM swings — the fresh-fork OSM moved 10× between fork days, which is what actually broke the borrow specs in CI — by sizing collateral swing-proof (25M SKY), deriving the liquidation staging from live vat spot/dust, and restoring the vat spot the capped-OSM spec mutates. Full stake e2e set green locally on a fresh fork (19/19); unit 1841/1841.

File: apps/webapp/src/test/e2e/utils/stakeV2.ts [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+) {
+  await stakeDeepLink(page, 'flow=open');
+  await expect(page.getByTestId('stake-takeover')).toBeVisible({ timeout: 15_000 });
+
+  await page.getByTestId('stake-takeover-stake-amount').fill(sky);
+
+  if (usds) {
+    await page.getByTestId('stake-takeover-borrow-card-toggle').click();
+    await page.getByTestId('stake-takeover-borrow-amount').fill(usds);
+  }
```
[the remark this answers is not in the document]
Comment (codewiththeflowxyz) [reply apps/webapp/src/test/e2e/utils/stakeV2.ts:74]: Confirmed and fixed in d902cb95a — though the root cause turned out not to be a timing race: forks inherit mainnet's live OSM for LSEV2-SKY-A, and it dropped 10× between fork days, which put the 30K-USDS dust floor at 14.4M SKY of collateral. Every spec staking 2.4M therefore left the borrow input *legitimately* disabled (`minCollateralNotMet`), which is exactly the disabled state in the CI logs.

Fixes: all borrow-involving specs now stake a swing-proof `BORROW_SPEC_SKY` (25M, documented against the floor math), `stageLiquidatedUrn` sizes its lock off the live vat spot/dust instead of a hardcoded amount, and `openStakePosition` additionally waits for the input to be enabled before filling (the debounced zero-stake simulation disables it briefly even in the healthy case). On the shared-fork isolation concern raised alongside this: the immediate state leak is closed — the capped-OSM spec now restores the vat spot it mutates in an `afterEach` — and moving the fork-mutating specs into their own Playwright project is queued as a follow-up. Full stake e2e set validated green locally on a fresh fork (19/19).

**Comment by codewiththeflowxyz** [comment]
Re: the debt-ceiling boundary finding in https://github.com/jetstreamgg/tarmac/pull/1710#issuecomment-4918597528 — confirmed and fixed in 3290bdce9.

Borrow validity now uses `<=` against `availableBorrowFromDebtCeiling` in both flows (`OpenPositionTakeover`, `ManagePositionTakeover`), so the exact headroom the 100% chip stages when the ceiling binds is valid and Confirm stays enabled; strictly above it still renders the debt-ceiling error. `<=` is safe here since the headroom is already computed against the 0.001%-padded total debt, so equality keeps a real buffer under the ceiling. Added boundary unit tests in both flows: 100% chip at a binding ceiling → no error + Confirm enabled; one unit above → error + disabled (the first fails against the old strict `<`).

File: apps/webapp/src/modules/stake/components/StakeTakeoverDelegateCard.tsx [hunk 2]
```diff
@@ -0,0 +1,159 @@
...
+                          `https://vote.sky.money/address/${delegate.id.toLowerCase()}`
+                        }
+                        target="_blank"
+                        rel="noopener noreferrer"
+                        onClick={event => event.stopPropagation()}
+                        aria-label={t`View delegate profile`}
+                        className="text-textSecondary hover:text-text"
+                      >
+                        <ExternalLink className="h-3.5 w-3.5" aria-hidden />
+                      </a>
```
[the remark this answers is not in the document]
Comment (codewiththeflowxyz) [reply apps/webapp/src/modules/stake/components/StakeTakeoverDelegateCard.tsx:102]: Fixed in 9a59ac43b. Added `lib/delegateProfileUrl.ts`: only `http:`/`https:` schemes pass through (parsed via `new URL`, so whitespace/control-char scheme smuggling is normalized before the check); anything else — `javascript:`, `data:`, protocol-relative, malformed — falls back to the delegate's vote.sky.money profile. Unit tests cover the bypass vectors.

File: apps/webapp/src/modules/stake/components/StakeManageBorrowCard.tsx [hunk 3]
```diff
@@ -0,0 +1,310 @@
...
+      if (maxRepayable === 0n) return;
+      const raw = percent === 100 ? maxRepayable : ((maxRepayable * BigInt(percent)) / 100n / WAD) * WAD;
+      // wipeAll only when the exact-max staging clears the full debt (M11).
+      onAmountChange(raw, percent === 100 && maxRepayable === existingDebt && existingDebt > 0n);
+      return;
+    }
+    if (maxBorrowable === 0n) return;
+    const raw = percent === 100 ? maxBorrowable : (maxBorrowable * BigInt(percent)) / 100n;
+    onAmountChange(percent === 100 ? raw : (raw / WAD) * WAD);
+  };
```
[the remark this answers is not in the document]
Comment (codewiththeflowxyz) [reply apps/webapp/src/modules/stake/components/StakeManageBorrowCard.tsx:123]: Fixed in 9a59ac43b. The slider's `setUsdsToWipe` path now mirrors the 100% chip's condition — `wipeAll` is staged when the dragged value equals the full debt and the balance covers it. Verified the failure mode first: a full-left drag lands on the exact debt (the `repayPercentage === repayablePercentage` branch returns it unrounded), and a plain wipe there strands sub-dust accrued interest (vat dust revert). Regression test drives the actual Radix slider via the Home key.

File: apps/webapp/src/modules/stake/components/PositionManageFlow.tsx [hunk 4]
```diff
@@ -0,0 +1,141 @@
...
+  onInitialSheetInitConsumed?: () => void;
+} = {}) {
+  const [searchParams, setSearchParams] = useAppSearchParams();
+  const urnIndex = parseUrnIndex(searchParams.get(QueryParams.UrnIndex));
+
+  const [view, setView] = useState<ManageView>(() => {
+    if (initialSheetInit) return { name: 'sheet', init: initialSheetInit };
+    const init = stakeTabInit(searchParams.get(QueryParams.StakeTab));
+    return init ? { name: 'sheet', init } : { name: 'details' };
+  });
```
[the remark this answers is not in the document]
Comment (codewiththeflowxyz) [reply apps/webapp/src/modules/stake/components/PositionManageFlow.tsx:76]: Fixed in 9a59ac43b. The `isLiquidatedStakePosition` guard now runs before the view dispatch, so both mount-time sheet paths (`stake_tab` deep link and a caller-staged `initialSheetInit`) land on the post-mortem for a barked urn. Added coverage for both paths; the recovery fall-through behavior (guard flips false after the SKY is freed) is unchanged.

File: apps/webapp/src/components/product/TakeoverShell.tsx [hunk 5]
```diff
@@ -0,0 +1,98 @@
...
+  return (
+    <div
+      role="dialog"
+      aria-modal="true"
+      data-testid={dataTestId}
+      // Same page-background recipe as the shell surface (shellLayoutClasses):
+      // `bg-background` is undefined in the dark scope (known token gap) and
+      // would leave the overlay transparent.
+      className="bg-app-background light:bg-blend-normal fixed inset-0 z-40 flex flex-col [background-color:#040434] bg-cover bg-center bg-no-repeat bg-blend-luminosity"
+    >
```
[the remark this answers is not in the document]
Comment (codewiththeflowxyz) [reply apps/webapp/src/components/product/TakeoverShell.tsx:98]: Fixed in 9a59ac43b. The shell now moves focus onto the dialog on open, wraps Tab/Shift+Tab at the edges, restores focus to the trigger on close, and names itself via `aria-labelledby` on the title. One deliberate choice: the Tab trap listens on the container rather than the document — the transaction modal (z-50) portals above this shell with its own Radix focus scope, and a document-level trap would yank focus back out of it. Tests cover initial focus, both wrap directions, restore, and the label link.

File: apps/webapp/src/modules/stake/components/StakeManageConfirmSummary.tsx [hunk 6]
```diff
@@ -0,0 +1,133 @@
...
+  usdsToWipe: bigint;
+  skyPriceUsd: number | null;
+  /** Set both to render the delegate From → To block. */
+  delegateFrom?: `0x${string}`;
+  delegateTo?: `0x${string}`;
+}) {
+  const skyUsd = (amount: bigint) =>
+    skyPriceUsd !== null ? Number(formatUnits(amount, 18)) * skyPriceUsd : null;
+  const hasAmounts = skyToLock > 0n || skyToFree > 0n || usdsToBorrow > 0n || usdsToWipe > 0n;
+  const showDelegate = !!delegateTo && !hasAmounts;
```
[the remark this answers is not in the document]
Comment (codewiththeflowxyz) [reply apps/webapp/src/modules/stake/components/StakeManageConfirmSummary.tsx]: Fixed in 9a59ac43b. Dropped the `!hasAmounts` suppression — a staged delegate change now previews its From → To block in mixed bundles too, matching the "Change delegate" step that was already listed. New test file pins delegate-only, mixed, and no-change cases.

File: apps/webapp/src/modules/stake/components/StakeTakeoverAmountField.tsx [hunk 7]
```diff
@@ -0,0 +1,100 @@
...
+              </button>
+            ))}
+          </div>
+        )}
+      </div>
+      {error && (
+        <span data-testid={`${dataTestId}-error`} className="text-error text-sm">
+          {error}
+        </span>
+      )}
```
[the remark this answers is not in the document]
Comment (codewiththeflowxyz) [reply apps/webapp/src/modules/stake/components/StakeTakeoverAmountField.tsx:99]: Fixed in 9a59ac43b. The error span now carries an id and the input references it via `aria-describedby` when an error is present (alongside the existing `aria-invalid`). Asserted in the insufficient-funds test.

File: apps/webapp/src/modules/stake/lib/manageRepay.test.ts [hunk 8]
```diff
@@ -0,0 +1,47 @@
...
+    // debt 40k, balance 5k → remaining 35k ≥ dust → repay balance 5k.
+    expect(calculateMaxRepayable({ debtValue: usds('40000'), dust: DUST, balance: usds('5000') })).toEqual(
+      usds('5000')
+    );
+    // debt 30,000.10, balance 29,999 → remaining ~1 < dust; maxWithoutDust ≈ 0.10
+    // and the balance covers it → 0.10.
+    expect(calculateMaxRepayable({ debtValue: usds('30000.1'), dust: DUST, balance: usds('29999') })).toEqual(
+      usds('0.1')
+    );
+  });
```
[the remark this answers is not in the document]
Comment (codewiththeflowxyz) [reply apps/webapp/src/modules/stake/lib/manageRepay.test.ts]: Fixed in 9a59ac43b. The mislabeled block was renamed to describe what it actually asserts, and a real "returns 0" test now exercises both unreachable branches: debt exactly at dust with a tiny balance (`maxRepayWithoutDust === 0`) and debt already below dust (negative gap).

File: apps/webapp/src/modules/stake/components/StakeAboutTab.tsx [hunk 9]
```diff
@@ -0,0 +1,123 @@
...
+ * Staking Engine promo card in the right rail. Read-only.
+ */
+export function StakeAboutTab() {
+  const banner = getBannerById(ABOUT_BANNER_ID);
+  const chainId = useChainId();
+  const contractHref = getEtherscanLink(
+    chainId,
+    stakeModuleAddress[chainId as keyof typeof stakeModuleAddress],
+    'address'
+  );
```
[the remark this answers is not in the document]
Comment (codewiththeflowxyz) [reply apps/webapp/src/modules/stake/components/StakeAboutTab.tsx]: Fixed in 9a59ac43b. The page has no hard chain gate, so the link now falls back to the mainnet contract (staking is mainnet-only) instead of rendering `/address/undefined` on chains without a module deployment.

File: apps/webapp/src/modules/stake/components/LiquidationPostMortemModal.tsx [hunk 10]
```diff
@@ -0,0 +1,376 @@
...
+        return {
+          contractAddress: reward.contractAddress,
+          symbol: reward.rewardSymbol,
+          amount: reward.claimBalance,
+          amountUsd: Number(formatUnits(reward.claimBalance, 18)) * price
+        };
+      });
+  }, [toClaim, prices]);
+
+  const claimableUsd = claims.reduce((total, claim) => total + claim.amountUsd, 0);
```
[the remark this answers is not in the document]
Comment (codewiththeflowxyz) [reply apps/webapp/src/modules/stake/components/LiquidationPostMortemModal.tsx:170]: Acknowledged but leaving as-is for this PR. The claimable set today is SKY/USDS/SPK, all covered by the price feed, and `$0`-for-missing-price is the convention across the module's USD aggregations — making just this one spot null-aware would be inconsistent. Worth a module-wide pass if a reward token ever ships ahead of its price-feed entry.

File: apps/webapp/src/modules/stake/hooks/useStakeManageLaunch.ts [hunk 11]
```diff
@@ -0,0 +1,316 @@
...
+  usdsToWipe: bigint;
+  wipeAll: boolean;
+  /** EFFECTIVE delegate: the staged selection, or the urn's current one (M12). */
+  selectedDelegate: `0x${string}` | undefined;
+  /** Form validity — gates the engine's prepare/simulation. */
+  enabled: boolean;
+  /** Reward contracts to bundle a getReward leg for, e.g. a liquidation-recovery claim. */
+  rewardContractsToClaim?: `0x${string}`[];
+  /** Display symbols aligned to `rewardContractsToClaim`, for step labels only. */
+  claimSymbols?: string[];
```
[the remark this answers is not in the document]
Comment (codewiththeflowxyz) [reply apps/webapp/src/modules/stake/hooks/useStakeManageLaunch.ts:82]: Deferring this one. It's an API-robustness concern rather than a live bug: the only caller today derives both `rewardContractsToClaim` and `claimSymbols` from the same `claims` memo, so they can't desync. Folding the symbol lookup into the hook is reasonable hardening if a second caller appears — noting it for a follow-up rather than expanding this PR.

File: apps/webapp/src/test/e2e/tests/capped-osm-unstake.spec.ts [hunk 12]
```diff
@@ -1,115 +1,63 @@
...
+    await isolatedPage.getByTestId('stake-manage-menu-withdraw').click();
+    await expect(isolatedPage.getByTestId('stake-manage-takeover')).toBeVisible();
+    await isolatedPage.getByTestId('stake-manage-stake-amount').fill('100000');
 
-    console.log('Continue button is disabled when capped OSM error occurs');
+    // Step 5: the capped OSM guard blocks the withdrawal.
+    await expect(
+      isolatedPage.getByText('Liquidation price is higher than the capped OSM SKY price')
+    ).toBeVisible({ timeout: 10_000 });
+    await expect(isolatedPage.getByTestId('stake-manage-confirm')).toBeDisabled();
```
[the remark this answers is not in the document]
Comment (codewiththeflowxyz) [reply apps/webapp/src/test/e2e/tests/capped-osm-unstake.spec.ts:75]: Deferring. The fixed sleep matches the harness's existing conventions (the suite's beforeEach uses fixed waits throughout) and the spec is green locally and in CI after the sizing fixes. Swapping sleeps for state-based waits is worth doing suite-wide in one pass rather than piecemeal in this spec.
